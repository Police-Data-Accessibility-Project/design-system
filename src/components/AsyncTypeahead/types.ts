export interface PdapAsyncTypeaheadProps<T> {
	id: string;
	placeholder?: string;
	items: T[] | undefined;
	formatItemForDisplay?: (item: T) => string;
	error?: string;
	position?: 'top' | 'bottom';
}
