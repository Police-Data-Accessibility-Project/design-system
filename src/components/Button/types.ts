export interface PdapButtonProps {
	intent?: 'primary' | 'secondary' | 'tertiary';
	isLoading?: boolean;
}
