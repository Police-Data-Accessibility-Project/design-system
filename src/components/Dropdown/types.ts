export enum PdapDropdownTriggerType {
	PRESS = 'press',
	HOVER = 'hover',
}

export interface PdapDropdownProps {
	defaultOpen?: boolean;
	disabled?: boolean;
	triggerOn?: PdapDropdownTriggerType;
}
