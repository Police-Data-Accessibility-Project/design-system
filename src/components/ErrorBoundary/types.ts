import { ComponentPublicInstance } from 'vue';

export interface PdapErrorEmitted {
	error: Error;
	vm: ComponentPublicInstance | null;
	info?: string;
}

export interface PdapErrorBoundaryProps {
	component: string;
	onError?: (
		error: Error,
		target?: ComponentPublicInstance | null | undefined,
		info?: string
	) => void;
	params?: Record<string, string>;
}
