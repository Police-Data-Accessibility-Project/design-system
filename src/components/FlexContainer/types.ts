export interface PdapFlexContainerProps {
	alignment?: 'center' | 'start';
	className?: string;
	component?: string;
}
