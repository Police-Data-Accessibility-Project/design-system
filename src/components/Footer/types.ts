import { FOOTER_LINK_ICONS } from './constants';

export interface PdapFooterSocialLinks {
	href?: string;
	path?: string;
	text: string;
	icon?: FooterIconName;
}
export type FooterIconName =
	(typeof FOOTER_LINK_ICONS)[keyof typeof FOOTER_LINK_ICONS];

export interface PdapFooterProps {
	fundraisingData: {
		raised: number;
		goal: number;
	};
}
