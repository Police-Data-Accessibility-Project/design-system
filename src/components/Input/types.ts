export enum PdapInputTypes {
	CHECKBOX = 'checkbox',
	TEXT = 'text',
}

export interface PdapInputBaseProps {
	errorMessage: string | undefined | null;
	id: string;
	label: string;
	name: string;
}

export interface PdapInputTextProps extends PdapInputBaseProps {
	placeholder?: string;
	type: PdapInputTypes.TEXT;
	value: string;
}

export interface PdapInputCheckboxProps extends PdapInputBaseProps {
	defaultChecked: boolean;
	type: PdapInputTypes.CHECKBOX;
	value: string;
}
