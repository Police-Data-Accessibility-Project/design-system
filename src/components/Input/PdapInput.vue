<template>
	<div
		:class="{
			'pdap-input': true,
			'pdap-input-error': error?.length >= 1,
		}"
	>
		<slot />
		<label class="pdap-input-label" :for="id">{{ label }}</label>
		<div v-if="error" :id="errorMessageId" class="pdap-input-error-message">
			{{ error }}
		</div>
	</div>
</template>

<script setup lang="ts">
// Globals
import { computed, watchEffect } from 'vue';

import { PdapInputBaseProps } from './types';

// Props
const props = defineProps<PdapInputBaseProps>();

const errorMessageId = computed(() => `pdap-${props.name}-input-error`);

watchEffect(() => {
	console.log(
		`Hello from PDAP Input ${props.name} rendered at ${window.location.pathname}`,
		{
			props,
		}
	);
});
</script>

<style>
@tailwind components;

@layer components {
	/* General input styles  */
	.pdap-input {
		@apply align-middle flex h-[max-content] gap-4 leading-normal mb-2 w-full;
	}

	.pdap-input input {
		@apply dark:bg-neutral-950 border-2 border-neutral-300 border-solid border-transparent px-3 py-2 text-[rgba(0,0,0)];
	}

	.pdap-input input::placeholder {
		@apply text-neutral-500;
	}

	.pdap-input input:focus,
	.pdap-input input:focus-within,
	.pdap-input input:focus-visible {
		@apply border-2 border-brand-blue-light border-solid outline-none;
	}

	.pdap-input label {
		@apply items-center flex basis-[max-content] justify-end p-3;
	}

	/* Error state */
	.pdap-input-error {
		@apply flex-wrap;
		row-gap: 0;
	}

	.pdap-input-error label {
		@apply justify-start;
	}

	.pdap-input-error input {
		@apply border-red-800 dark:border-red-300;
	}

	.pdap-input-error-message {
		@apply items-center justify-start basis-full flex-shrink flex bg-red-300 text-red-800 p-2 text-sm;
	}

	/* Specific inputs */

	/* Input - text */
	.pdap-input input[type='text'] {
		@apply h-12 text-lg;
	}

	/* Input - checkbox */
	.pdap-input input[type='checkbox'] {
		@apply w-6 accent-brand-gold;
	}
}
</style>
./Checkbox
