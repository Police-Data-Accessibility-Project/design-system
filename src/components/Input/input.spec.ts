import PdapInput from './PdapInput.vue';

// Utils
import { mount } from '@vue/test-utils';
import { describe, expect, test } from 'vitest';
import { PdapInputTypes } from './types';
import { nextTick } from 'vue';

const baseCheckbox = {
	props: {
		error: undefined,
		id: 'test-checked',
		label: 'test this checkbox input',
		name: 'testCheckbox',
		type: PdapInputTypes.CHECKBOX,
		defaultChecked: false,
	},
};

const baseText = {
	props: {
		error: undefined,
		id: 'test-text',
		label: 'test this text input',
		name: 'testText',
		placeholder: 'This is testing testing',
		type: PdapInputTypes.TEXT,
		value: '',
	},
};

describe('Input component', () => {
	test('Renders text input in okay state', () => {
		const wrapper = mount(PdapInput, baseText);

		expect(
			wrapper.find('.pdap-input').find('input[type="text"]').exists()
		).toBe(true);
		expect(wrapper.find('.pdap-input-error-message').exists()).toBe(false);
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Renders text input in error state', () => {
		const wrapper = mount(PdapInput, {
			...baseText,
			props: { ...baseText.props, error: 'error message' },
		});

		expect(
			wrapper.find('.pdap-input').find('input[type="text"]').exists()
		).toBe(true);
		expect(wrapper.find('.pdap-input-error-message').exists()).toBe(true);
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Renders password input in okay state', () => {
		const wrapper = mount(PdapInput, {...baseText, props: {...baseText.props, type: PdapInputTypes.PASSWORD}});

		expect(
			wrapper.find('.pdap-input').find('input[type="password"]').exists()
		).toBe(true);
		expect(wrapper.find('.pdap-input-error-message').exists()).toBe(false);
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Renders password input in error state', () => {
		const wrapper = mount(PdapInput, {
			...baseText,
			props: { ...baseText.props, error: 'error message', type: PdapInputTypes.PASSWORD},
		});

		expect(
			wrapper.find('.pdap-input').find('input[type="password"]').exists()
		).toBe(true);
		expect(wrapper.find('.pdap-input-error-message').exists()).toBe(true);
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Renders checkbox input in okay state', () => {
		const wrapper = mount(PdapInput, baseCheckbox);

		expect(
			wrapper.find('.pdap-input').find('input[type="checkbox"]').exists()
		).toBe(true);
		expect(wrapper.find('.pdap-input-error-message').exists()).toBe(false);
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Renders checkbox input in error state', () => {
		const wrapper = mount(PdapInput, {
			...baseCheckbox,
			props: { ...baseCheckbox.props, error: 'error message' },
		});

		expect(
			wrapper.find('.pdap-input').find('input[type="checkbox"]').exists()
		).toBe(true);
		expect(wrapper.find('.pdap-input-error-message').exists()).toBe(true);
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Text input accepts user input', async () => {
		const wrapper = mount(PdapInput, baseText);

		const input = wrapper.find('input[type="text"]');
		input.setValue('nonsense');
		await nextTick();

		expect((input.element as HTMLInputElement).value).toBe('nonsense');
	});

	test('Checkbox input accepts user input', async () => {
		const wrapper = mount(PdapInput, baseCheckbox);

		const input = wrapper.find('input[type="checkbox"]');
		await nextTick();

		// First assert falsy value
		expect((input.element as HTMLInputElement).checked).toBe(false);

		// Set checkbox value
		input.trigger('change');
		await nextTick();

		expect((input.element as HTMLInputElement).checked).toBe(true);
	});
});
