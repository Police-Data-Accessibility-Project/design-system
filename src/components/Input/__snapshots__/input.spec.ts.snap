// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input component > Renders checkbox input in error state 1`] = `
<div class="pdap-input pdap-input-checkbox pdap-input-error">
  <input class="pdap-input-checkbox" id="test-checked" name="testCheckbox" type="checkbox">
  <div class="pdap-input-error-message" id="pdap-testCheckbox-input-error">error message</div>
  <label class="pdap-input-label" for="test-checked">test this checkbox input</label>
</div>
`;

exports[`Input component > Renders checkbox input in okay state 1`] = `
<div class="pdap-input pdap-input-checkbox">
  <input class="pdap-input-checkbox" id="test-checked" name="testCheckbox" type="checkbox">
  <!--v-if-->
  <label class="pdap-input-label" for="test-checked">test this checkbox input</label>
</div>
`;

exports[`Input component > Renders password input in error state 1`] = `
<div class="pdap-input pdap-input-password pdap-input-error">
  <input id="test-text" name="testText" placeholder="This is testing testing" type="password">
  <div class="pdap-input-error-message" id="pdap-testText-input-error">error message</div>
  <label class="pdap-input-label" for="test-text">test this text input</label>
</div>
`;

exports[`Input component > Renders password input in okay state 1`] = `
<div class="pdap-input pdap-input-password">
  <input id="test-text" name="testText" placeholder="This is testing testing" type="password">
  <!--v-if-->
  <label class="pdap-input-label" for="test-text">test this text input</label>
</div>
`;

exports[`Input component > Renders text input in error state 1`] = `
<div class="pdap-input pdap-input-text pdap-input-error">
  <input id="test-text" name="testText" placeholder="This is testing testing" type="text">
  <div class="pdap-input-error-message" id="pdap-testText-input-error">error message</div>
  <label class="pdap-input-label" for="test-text">test this text input</label>
</div>
`;

exports[`Input component > Renders text input in okay state 1`] = `
<div class="pdap-input pdap-input-text">
  <input id="test-text" name="testText" placeholder="This is testing testing" type="text">
  <!--v-if-->
  <label class="pdap-input-label" for="test-text">test this text input</label>
</div>
`;
